Sources/FluentData/FluentDataContext.swift

```diff
@@ -331,77 +330,168 @@ extension FluentDataContext: DatabaseStateTracker {
+                    }
+
+                default:
+                    return .sqlite(.file(iCloudFilePath.path))
+                }
+            } else {
+                return try fallbackToLocalContainer()
+            }
 
         case .memory:
             return .sqlite(.memory)
         }
     }
 
-    static var filePath: URL? {
+    static func filePath() throws -> URL? {
         switch Self.persistence {
         case .bundle(let bundle, let name):
             if let path = bundle.url(forResource: name, withExtension: "sqlite") {
                 return path
             } else {
-                fatalError("Database \(name).sqlite not found in \(bundle.bundlePath).")
+                throw FluentDataContextError.bundledDatabaseNotFound
             }
 
         case .file(let name):
-            guard let folder = try? FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask, appropriateFor: nil, create: true) else {
-                fatalError("Unknown to store database `\(name)` (1)")
+            guard let applicationSupportDirectory = try? FileManager.default.url(
+                for: .applicationSupportDirectory,
+                in: .userDomainMask,
+                appropriateFor: nil,
+                create: true
+            ) else {
+                throw FluentDataContextError.unknownPathToDatabaseFile
             }
 
             guard let urlSafePersistenceName = name.addingPercentEncoding(
                 withAllowedCharacters: CharacterSet.urlPathAllowed.subtracting(CharacterSet.symbols).subtracting(CharacterSet.newlines)
             ) else {
-                fatalError("Unknown to store database `\(name)` (2)")
+                throw FluentDataContextError.invalidDatabaseName
             }
 
-            let filePath = folder.appendingPathComponent("\(urlSafePersistenceName).sqlite")
+            let filePath = applicationSupportDirectory.appendingPathComponent("\(urlSafePersistenceName).sqlite")
+            return filePath
+
+        case .iCloud(let container, let name):
+            guard let iCloudFolder = FileManager.default.url(forUbiquityContainerIdentifier: container) else {
+                return try Self.localFallbackFilePathUnsafe()
+            }
+
+            guard let urlSafePersistenceName = name.addingPercentEncoding(
+                withAllowedCharacters: CharacterSet.urlPathAllowed.subtracting(CharacterSet.symbols).subtracting(CharacterSet.newlines)
+            ) else {
+                throw FluentDataContextError.invalidDatabaseName
+            }
+
+            let filePath = iCloudFolder.appendingPathComponent("\(urlSafePersistenceName).sqlite")
             return filePath
 
         case .memory:
             return nil
         }
     }
 
-    static var removableFilePath: URL? {
+    static func filePathUnsafe() throws -> URL {
+        guard let path = try filePath() else {
+            fatalError("filePathUnsafe called for an unsupported persistence type")
```

Comment:
This func can throw, maybe u can throw here instead of fatalError?

Reply:
Except if it's an internal logic 🤔

Reply:
It's internal logic. It should never happen, we should never merge code that would trigger this line